### PR TITLE
Room accessible flag & DB migration refactor

### PR DESCRIPTION
**Pull request by jagerman**
1. `sogs.db` was getting messy with all the migration code; this moves it all into multiple files in `sogs.migrations.*`, where each migration submodule has a `migrate(conn)` function to perform a migration.

   Aside from moving all the migrations, this also renames/updates the existing `sogs.migrate01x` to `sogs.migrations.v_0_1_x` for consistency.

2.  This adds a new 'accessible' permission bit that allows restricting all metadata room info.  This allows you to distinguish between rooms that should be entirely secret (i.e. that random users can't learn anything about):

        read=False, accessible=False

    and rooms where users can get metadata but not read messages:

        read=False, accessible=True

### CODE REVIEW

File: sogs/migrations/__init__.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    ):
+        changes = False
+        with db.transaction(conn):
+            changes = migration.migrate(conn)
+        if changes:
+            db.metadata.clear()
+            db.metadata.reflect(bind=db.engine, views=True)
+
+
+#        migrate_v01x,
```
Comment (majestrate) [inline sogs/migrations/__init__.py]: what are these comments for?

File: sogs/migrations/import_hacks.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+            db.HAVE_FILE_ID_HACKS = True
+
+    try:
+        rows = conn.execute(
+            "SELECT room, old_message_id_max, message_id_offset FROM room_import_hacks"
+        )
+        for (room, id_max, offset) in rows:
+            db.ROOM_IMPORT_HACKS[room] = (id_max, offset)
+    except Exception:
+        pass
```
Comment (majestrate) [inline sogs/migrations/import_hacks.py:39]: do we really want a silent fail here?

File: sogs/migrations/new_columns.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+def migrate(conn):
+    """
+    New columns that might need to be added that don't require more complex migrations beyond simply
+    adding the column.
+    """
+
+    from .. import db
+
+    new_table_cols = {
```
Comment (majestrate) [inline sogs/migrations/new_columns.py:12]: why a set instead of a list?

File: sogs/model/room.py [hunk 4]
```diff
@@ -1400,15 +1447,15 @@ def get_rooms_with_permission(
...
-def get_readable_rooms(user: Optional[User] = None):
+def get_accessible_rooms(user: Optional[User] = None):
     """
-    Get a list of rooms that a user can access; if user is None then return all publicly readable
+    Get a list of rooms that a user can access; if user is None then return all publicly accessible
     rooms.
     """
     if user is None:
-        result = query("SELECT * FROM rooms WHERE read ORDER BY token")
+        result = query("SELECT * FROM rooms WHERE (read OR accessible) ORDER BY token")
```
Comment (majestrate) [inline sogs/model/room.py:1456]: should this be `read AND accessible` ?

File: sogs/routes/auth.py [hunk 5]
```diff
@@ -135,6 +135,36 @@ def require_mod(room, *, admin=False):
...
 
+def accessible_required(f):
+    """Decorator for an endpoint that requires a user have accessible or read permission in the
+    given room.  The function must take a `room` argument by name, as is typically used with flask
+    endpoints with a `<Room:room>` argument."""
+
+    @wraps(f)
+    def required_accessible_wrapper(*args, room, **kwargs):
+        if not room.check_accessible(g.user):
+            abort_with_reason(http.FORBIDDEN, "This room is not accessible", warn=False)
```
Comment (majestrate) [inline sogs/routes/auth.py]: this can leak info between rooms not existing and rooms not being accessible.

File: sogs/model/room.py (hunk 4 above)
Comment (jagerman) [reply sogs/model/room.py:1456]: I don't think so.  `accessible` only has meaning when `read` is false; when `read` is true we don't care what `accessible` is (or to put it another way: `read` implies a room being accessible).  So a room is accessible to you if: `(read or (not read and accessible))` which simplifies to `read or accessible`.

File: sogs/routes/auth.py (hunk 5 above)
Comment (jagerman) [reply sogs/routes/auth.py]: Yes, but does that actually matter?

File: sogs/migrations/__init__.py (hunk 1 above)
Comment (jagerman) [reply sogs/migrations/__init__.py]: Oops.  Leftover scaffolding from when I was moving things.  Removed.

File: sogs/migrations/import_hacks.py (hunk 2 above)
Comment (jagerman) [reply sogs/migrations/import_hacks.py:39]: Yes: if the table doesn't exist the query fails and throws, which is fine: we have no hack values to worry about.

File: sogs/migrations/new_columns.py (hunk 3 above)
Comment (jagerman) [reply sogs/migrations/new_columns.py:12]: Seems a bit cleaner to declare, and we don't care about the order these get applied.

File: sogs/routes/auth.py (hunk 5 above)
Comment (majestrate) [reply sogs/routes/auth.py]: yes, the point of this flag was to prevent this type of info leak.

Comment (jagerman) [reply sogs/routes/auth.py]: Updated to 404 instead.

**Review by majestrate — APPROVED** [review]

**Pull request merged**